### PR TITLE
feat: build c-toxcore with the zig build system

### PR DESCRIPTION
**Pull request by freylax**
feat: Building c-toxcore with zig.
The dependencies are in build.zig.zon 
The build.zig contains the build details.
Build the library: zig build install 
Allows cross compilation by specifying the target with eg -Dtarget=x86_64-windows-gnu
Google test and google mock test files are also build.

- - -
This change is [Reviewable](https://reviewable.io/reviews/TokTok/c-toxcore/2548)

### CODE REVIEW

**Comment by nurupo** [comment]
Toxcore already uses  3 build systems: autotools, bazel and cmake. A normal project typically uses just one. There should be a very compelling reason for adding yet another build system. I'm all ears.

**Comment by freylax** [comment]
The reason to have c-toxcore build by zig is that this is very good includable for zig. We can write easy an wrapper in zig or we use c-toxcore as reference test for an zig toxcore implementation. Libsodium by the way does use zig as build system among others.

**Comment by nurupo** [comment]
>The reason to have c-toxcore build by zig is that this is very good includable for zig.

If we go down this rabbit hole, next we will be adding build systems for Nim, Crystal, Go, Java, Python, Haskell, Rust, etc., and will end up with 20+ build systems because "this is very good includable for X".

>We can write easy an wrapper in zig or we use c-toxcore as reference test for an zig toxcore implementation.

There is already a wrapper being written for Zig, https://github.com/TokTok/zig-toxcore-c, by you, in fact. Doesn't Zig have a way to link against a C library that was already built with autotools/bazel/cmake? There shouldn't be a need to add a Zig build system for that.

Is someone re-implementing toxcore in Zig? Could you link that project? Also, I don't see how having toxcore being built using a Zig build system would be any different than building it using autotools/bazel/cmake, and what role does a build system even have in testing the resulting toxcore binary against another toxcore implementation? One would typically write a separate testing program for that, independent of the build system, instead of coding it as a part of a build system.

---

There should be a very compelling reason for adding yet another build system and I have yet to see any compelling reason being mentioned.

Toxcore already uses 3 build systems: autotools, bazel and cmake, which I think is already a bit too much. Having to maintain 4 of them doesn't help the situation. Each build system is a maintenance burden and it quickly adds up, to the point that changing something non-trivial in the build process of toxcore would require knowing autotools, bazel, cmake and now zig, which I'm not sure if we even have a person who knows all 4 of those. We shouldn't be adding a yet another build system without a good reason.

If a Zig build system brings something desirable to toxcore that none of the other build systems toxcore uses do, then it might be worth considering. But if it's no better than the 3 build systems toxcore already uses, then what is the point?

Then there is also the question if any of the regular contributors know Zig and if any of them would be willing to maintain this Zig build system. If it goes unmaintained once it gets merged, then what is the point?

**Comment by nurupo** [comment]
To me this PR look like someone wrote a build system in their favorite programming language just for the fun of it, not because the build system would benefit the project being built, and is now trying to push it upstream. Similar to if someone tried to contribute a SCons build script for toxcore just because they are very passionate about Python and SCons. Good for you, but it doesn't really make sense for toxcore to accept such a PR.

File: build.zig
Comment (iphydf) [inline build.zig:49]: Can we extract this list from the cmake code? I'd like to address nurupo's very valid concern by making this build system as easy to maintain as any other ci check. E.g. the conanfile is kind of a build system wrapper around cmake. If we can make the zig one a wrapper too, I think it would be less of a concern.

**Review by iphydf — CHANGES_REQUESTED** [review]

**Comment by nurupo** [comment]
This PR got some thumb-ups, so I assume those who let them consider the Zig build system to be somehow desirable, yet somehow no one managed to provide a reason of why we want toxcore to support a 4th build system, either on here on github or in the chat.

Since I don't know anything about Zig, I done some reading to learn more about it.

<details><summary>(A couple of paragraphs on the Zig C/C++ (cross-)compiler, which I found interesting and which I initially thought was the reason for the `build.zig` inclusion, but later realized that not being the case, so they can be skipped.)
</summary>

---

During the reading, I have stumbled upon this post: [`zig cc`: a Powerful Drop-In Replacement for GCC/Clang](https://andrewkelley.me/post/zig-cc-powerful-drop-in-replacement-gcc-clang.html) by the Zig programming language author/designer. It turns out that a distribution of Zig includes a clang-based C compiler that can be invoked as `zig cc`, accepting clang options as it's essentially a wrapper for clang. However, unlike your typical clang, Zig bundles Linux kernel headers, and binaries and headers of: glibc, musl and mingw-w64, for many different architectures: aarch64, arm, i386, mips, powerpc, riscv, x86_64, etc., and it does so for many different versions of glibc and musl libc. This allows `zig cc` to easily cross-compile any C project to one of numerous ABIs it supports (over 40), for example `riscv64-linux-musl`, `aarch64-windows-gnu` or `x86_64-windows-gnu`, as well as target a specific versions of the libc, like `x86_64-linux-gnu.2.25` for glibc 2.25. So using `zig cc` saves a lot of trouble of downloading/building separate cross-compilers for every ABI, having the Linux headers, glibc, musl libc, mingw-w64, etc. for the ABI you want to target installed, etc. -- all you need is just to download zig, run `zig cc foo.c -o foo.exe --target x86_64-windows-gnu` and you get your C code cross-compiled to Windows without having anything extra installed besides zig as zig is fully statically compiled and includes everything needed for the cross-compilation. You can even run it on Windows to create Linux binaries. It also works with C++ code via `zig c++`. It's pretty much a portable cross-compiler for C/C++ code, and it's also very light as it does some clever deduplication to reduce all the Linux headers, glibc, musl libc and mingw-w64, (and Zig itself also ships mostly as source code that is compiled and cached at runtime), to the point that its tarball takes 45 MiB according to the post.

So Zig is great for cross-compiling C/C++ code. One can use the Zig clang-based compiler by setting `CC="zig cc -target foo-bar-abi"` `CXX="zig c++ -target foo-bar-abi"` and invoking the existing autotools/bazel/cmake build system. If the project has dependencies, and toxcore does: sodium, opus, vpx, config, cmp -- those, and all of their dependencies, also need to be cross-compiled the same way, i.e. using the Zig C compiler (this is not specific to Zig, that's just how cross-compiling works).

---
</details>

While the Zig C compiler is a neat feature of Zig, using it doesn't necessitate the addition of `build.zig`, i.e. it doesn't require you to use the Zig build system in any way, it works just fine with the existing build systems toxcore uses. The only reasons to add a `build.zig` to a C project that I see are:

1. If you simply want to use the Zig build system to build your C project (perhaps that's your only build system instead of the 4th? :D)
2. If you want other projects using the Zig build system, be it C, C++ or Zig-native projects, to be able to easily add your project as a dependency to their Zig build system. ([Since many C/C++ projects don't provide a `build.zig`, developers using the Zig build system for their project with C/C++ dependencies tend to maintain forks of those C/C++ dependencies, e.g. ffmpeg, harfbuzz, freetype, etc. with `build.zig` added as the build system](https://devlog.hexops.com/2023/mach-ecosystem-c-libraries/)).

So this PR -- the addition of the Zig build system -- doesn't provide any benefits to toxcore over the existing autotools/bazel/cmake, this PR is more about providing benefits to the downstream projects that use the Zig build system and want to use toxcore. Also, this PR links only against sodium, cmp, pthreads and gtest, it doesn't link against opus, vpx and config, so it build only the base toxcore library, not toxav or tox-bootrstrap, so there is more work that needs to be done to build the toxcore project in its entirety using the Zig build system. What is more, opus, vpx and config don't provide `build.zig`, so we would either have to write our own Zig build systems for those, or use someone else's if someone else has already done the work of writing `build.zig`, i.e. instead of pulling the project from https://github.com/xiph/opus, we would be pulling it from some fork https://github.com/some-random-person/opus-zigbuild, which would obviously need to be audited and pinned to the audited version, and every time we version bump the dependency, it would need to be re-audited again... In fact, this PR already does this with winpthreads and gtest -- it builds them from the forks https://github.com/kassane/winpthreads-zigbuild and https://github.com/nodecum/gtest-zig instead of the original trusted repositories/sources. (I wonder if those were audited?)

Accepting this PR would result in a lot of work with getting the rest of toxcore dependencies Zig-ified, and then continuously making sure that our `build.zig` match what the upstream's build system does, or continuously auditing someone else's `build.zig` and the accompanied patched source code (I noticed winpthreads-zigbuild not only adds the `build.zig` but also patches the source code...). If you would be maintaining this `build.zig` in the [zig-toxcore-c](https://github.com/TokTok/zig-toxcore-c) anyway, I can see why you would want to have it reside in this repository instead, but this dependency stuff is just a bit too crazy.

If some Tox/TokTok developer is willing to take the responsibility for maintaining it, and do it properly -- actually audit the`buid.zig` forks used for dependencies as they are a lot less trusted than the original projects, then I guess fine. In the worst case we could just remove it if it gets unmaintained. But personally I'm not very thrilled about this PR.

**Comment by JFreegman** [comment]
@nurupo I actually agree with you despite my thumbs-up. We just got rid of autotools because of how cumbersome maintaining 3 build systems was. I don't really see how this benefits the project at the moment.

**Comment by freylax** [comment]
Thanks @nurupo for your careful analysis. I do understand the concerns you have. In fact it is not necessary to have the build.zig  in the c-toxcore repo. Like I did for gtest one can have c-toxcore as dependency in a separate repository. Also linking c-toxcore directly could be possible, but then maybe loosing the seamless cross compilation properties. The motivation of this PR was to move the dependency from zig-toxcore-c to TokTok.

File: build.zig
Comment (freylax) [reply build.zig:49]: This means parsing of Makefile.inc and use the files in the libtoxcore_la_SOURCES variable. 
And we drop the gtest stuff for now? For the downstream zig use they are not important anyway.
But then maybe we should use a fixed zig version for CI because a new zig could require the change of build.zig and this would be a bourdon for other maintainers.

**Pull request closed without merging**
